File: app/src/main/java/com/testapp/testapp/presenter/VenueDetailsPresenter.java
package com.testapp.testapp.presenter;

import android.support.annotation.NonNull;

import com.testapp.testapp.model.entity.Venue;
import com.testapp.testapp.model.entity.response.Response;
import com.testapp.testapp.model.entity.response.ResponseVenue;
import com.testapp.testapp.model.rest.RestApiManager;
import com.testapp.testapp.presenter.utils.RequestParametersHolder;
import com.testapp.testapp.view.CustomItemView;

import retrofit2.Call;
import retrofit2.Callback;

/**
 * Created by troll on 07.11.2017.
 */

public class VenueDetailsPresenter implements Presenter, Callback<Response<ResponseVenue>> {

    private CustomItemView<Venue> view;
    private String venueId;

    public VenueDetailsPresenter(CustomItemView<Venue> view, String venueId) {
        this.view = view;
        this.venueId = venueId;
    }

    @Override
    public void getResponse() {
        Call<Response<ResponseVenue>> call = new RestApiManager()
                .getVenueApi()
                .venueDetails(venueId, RequestParametersHolder.getInstance().getCommonRequestParams());
        call.enqueue(this);
        view.onStartRequest();
    }

    @Override
    public void onResponse(@NonNull Call<Response<ResponseVenue>> call, @NonNull retrofit2.Response<Response<ResponseVenue>> response) {
        Response<ResponseVenue> commonResponse = response.body();
        if (commonResponse != null) {
            Venue responseVenue = commonResponse.getResponse().getVenue();
            responseVenue.setPrimaryCategory();
            view.onSuccessResponse(responseVenue);
        }
        view.onEndRequest();
    }

    @Override
    public void onFailure(@NonNull Call<Response<ResponseVenue>> call, @NonNull Throwable t) {
        view.onFailureRequest();
    }
}

File: app/src/main/java/com/testapp/testapp/kotlin/presenter/VenueDetailsPresenter.kt
package com.testapp.testapp.kotlin.presenter


import com.testapp.testapp.kotlin.model.entity.Venue
import com.testapp.testapp.kotlin.model.entity.response.Response
import com.testapp.testapp.kotlin.model.entity.response.ResponseVenue
import com.testapp.testapp.kotlin.model.rest.RestApiManager
import com.testapp.testapp.kotlin.presenter.utils.RequestParametersHolder
import com.testapp.testapp.kotlin.view.CustomView
import retrofit2.Call
import retrofit2.Callback

class VenueDetailsPresenter(val view: CustomView<Venue?>, val venueId: String): Presenter, Callback<Response<ResponseVenue>> {
    override fun getResponse() {
        val restApiManager = RestApiManager.instance
        val responseCall = restApiManager.getVenueApi()
                ?.venueDetails(venueId, RequestParametersHolder.instance.requestParameters)
        responseCall?.enqueue(this)
        view.onStartRequest()
    }

    override fun onFailure(call: Call<Response<ResponseVenue>>?, t: Throwable?) {
        view.onFailureRequest()
    }

    override fun onResponse(call: Call<Response<ResponseVenue>>?, response: retrofit2.Response<Response<ResponseVenue>>?) {
        val commonResponse = response?.body()
        val responseVenue = commonResponse?.response?.venue
        responseVenue?.setPrimaryCategory()
        view.onSuccessResponse(responseVenue)
        view.onEndRequest()
    }

}